cli/lsp/testing/execution.rs

```diff
@@ -323,133 +366,259 @@ impl TestRun {
-                let description = tests.read().get(&id).unwrap().clone();
-                match &result {
-                  test::TestResult::Ok => summary.passed += 1,
-                  test::TestResult::Ignored => summary.ignored += 1,
-                  test::TestResult::Failed(error) => {
-                    summary.failed += 1;
-                    summary
-                      .failures
-                      .push(((&description).into(), error.clone()));
-                  }
-                  test::TestResult::Cancelled => {
-                    summary.failed += 1;
-                  }
-                }
-                reporter.report_result(&description, &result, elapsed);
+      while let Some((_, event)) = receiver.recv().await {
+        test_event_handler.handle_event(event).await;
+      }
+
+      test_event_handler.finish(start_time.elapsed())?;
+
+      Result::<_, AnyError>::Ok(())
+    });
+
+    let (join_results, result) = future::join(join_stream, handler).await;
+
+    // propagate any errors
+    for join_result in join_results {
+      join_result??;
+    }
+
+    result??;
+
+    Ok(())
+  }
+
+  async fn run_tests_with_watch(
+    &self,
+    client: &Client,
+    maybe_root_uri: Option<&ModuleSpecifier>,
+  ) -> Result<(), AnyError> {
+    let args = self.get_args();
+    lsp_log!("Executing test run with arguments: {}", args.join(" "));
+    let flags = Arc::new(flags_from_vec(
+      args.into_iter().map(|s| From::from(s.as_ref())).collect(),
+    )?);
+
+    let (concurrent_jobs, fail_fast) = {
+      let factory = CliFactory::from_flags(flags.clone());
+      let cli_options = factory.cli_options()?;
+      if let DenoSubcommand::Test(test_flags) = cli_options.sub_command() {
+        let concurrent_jobs = test_flags
+          .concurrent_jobs
+          .unwrap_or_else(|| NonZeroUsize::new(1).unwrap())
+          .into();
+
+        // TODO(mmastrac): Temporarily limit concurrency in windows testing to avoid named pipe issue:
+        // *** Unexpected server pipe failure '"\\\\.\\pipe\\deno_pipe_e30f45c9df61b1e4.1198.222\\0"': 3
+        // This is likely because we're hitting some sort of invisible resource limit
+        // This limit is both in cli/lsp/testing/execution.rs and cli/tools/test/mod.rs
+        #[cfg(windows)]
+        let concurrent_jobs = std::cmp::min(concurrent_jobs, 4);
+        (concurrent_jobs, test_flags.fail_fast)
+      } else {
+        unreachable!("Should always be Test subcommand.");
+      }
+    };
+    let (test_run_event_sender, mut receiver) =
+      tokio::sync::mpsc::unbounded_channel::<TestRunEvent>();
+
+    let mut first_run_token = Some(());
+    lsp_log!("Starting the setup");
+    let join_handle = file_watcher::watch_recv(
+      flags.clone(),
+      file_watcher::PrintConfig::new("Test", false),
+      file_watcher::WatcherRestartMode::Automatic,
+      self.token.clone(),
+      move |flags, watcher_communicator, changed_paths| {
```

Comment:
I'm not entirely sure which files I should pass to the watcher_communicator, and what to do with the changed_paths. I did look at the other places where the file watcher is used, but there the file patterns are part of the CLI arguments. 

Meanwhile here, the list of included files are set through the included and excluded tests.